components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/device/response/DeviceFlowResponseTypeHandler.java

```diff
@@ -56,10 +65,38 @@ public OAuth2AuthorizeRespDTO issue(OAuthAuthzReqMessageContext oauthAuthzMsgCtx
         String userCode = authzReqDTO.getNonce();
         DeviceFlowPersistenceFactory.getInstance().getDeviceFlowDAO().setAuthzUserAndStatus(userCode,
                 Constants.AUTHORIZED, authenticatedUser);
-        respDTO.setCallbackURI(authzReqDTO.getCallbackUrl());
+        OAuthAppDO oAuthAppDO = (OAuthAppDO) oauthAuthzMsgCtx.getProperty("OAuthAppDO");
+        String redirectionURI = getRedirectionURI(oAuthAppDO.getApplicationName(),
+                oauthAuthzMsgCtx.getAuthorizationReqDTO().getTenantDomain());
+        respDTO.setCallbackURI(redirectionURI);
         return respDTO;
     }
 
+    /**
+     * This method is used to generate the device flow authentication completed page URI.
+     *
+     * @param appName       Service provider name.
+     * @param tenantDomain  Tenant domain.
+     * @return Redirection URI
+     */
+    private String getRedirectionURI(String appName, String tenantDomain) throws IdentityOAuth2Exception {
+
+        try {
+            String pageURI = ServiceURLBuilder.create().addPath(Constants.DEVICE_SUCCESS_ENDPOINT_PATH)
+                    .build().getAbsolutePublicURL();
+            URIBuilder uriBuilder = new URIBuilder(pageURI);
+            uriBuilder.addParameter(Constants.APP_NAME, appName);
+            if (!IdentityTenantUtil.isTenantQualifiedUrlsEnabled() && isNotSuperTenant(tenantDomain)) {
```

Comment:
I think we have a util method available to these checks and build the uri.
@mefarazath @chamathns